### PR TITLE
Track welcome tutorial component in local storage

### LINKED ISSUE #147: Track Welcome Tutorial in localstorage
We should keep track of the welcome tutorial so we don't keep showing it every time someone loads the map.

We can track the current date in localstorage so that we can show the tutorial accordingly if someone hasn't refreshed the page in like a month or two.

### PR DESCRIPTION
**Pull request by hanlinc27**
Store date in local storage so that the welcome tutorial only shows when first entered once every 20 days, tested it with 15 second timeout. 
Tutorial will always show when launched from help button.
Example (15 seconds)
welcome-tutorial-demo

Closes #147

### CODE REVIEW

File: ui/src/pages/ShoeMap.tsx [hunk 1]
```diff
@@ -90,7 +90,33 @@ export const ShoeMap: React.FC = () => {
...
+    const item = {
+      value: "value",
+      expiry: new Date().getTime(),
+    };
+    localStorage.setItem(key, JSON.stringify(item));
+  };
+
+  const compareTimestamp = () => {
+    const oldTimestampStr = localStorage.getItem(key);
+    if (!oldTimestampStr) {
```
Comment (hanlinc27) [inline ui/src/pages/ShoeMap.tsx:107]: If it's the user's first time on the website

File: ui/src/styles/typography.ts [hunk 2]
```diff
@@ -54,7 +54,7 @@ export const CardDescriptionText = styled.p`
   font-family: Poppins;
   font-weight: normal;
   line-height: 24px;
-  text-align: left;
+  text-align: justified;
```
Comment (hanlinc27) [inline ui/src/styles/typography.ts:57]: This is unrelated, just a styling fix from Figma.

File: ui/src/pages/ShoeMap.tsx [hunk 3]
```diff
@@ -89,8 +89,34 @@ const StyledLogo = styled.div`
 export const ShoeMap: React.FC = () => {
   const zoom = 4;
   const minZoom = 4;
-  const maxZoom = 18;
+  const maxZoom = 12;
```
Comment (hanlinc27) [inline ui/src/pages/ShoeMap.tsx:100]: re: Amanda's message

File: ui/src/pages/ShoeMap.tsx [hunk 4]
```diff
@@ -89,8 +89,34 @@ const StyledLogo = styled.div`
...
+
+  const setTimeStamp = () => {
+    const item = {
+      value: "value",
+      expiry: new Date().getTime(),
+    };
+    localStorage.setItem(key, JSON.stringify(item));
+  };
+
+  const compareTimestamp = () => {
```
Comment (AbhiPrasad) [inline ui/src/pages/ShoeMap.tsx]: Let's move this func and `setTimeStamp` out of the component

File: ui/src/pages/ShoeMap.tsx [hunk 5]
```diff
@@ -89,8 +89,34 @@ const StyledLogo = styled.div`
...
+      return TutorialState.First;
+    }
+    const oldTimestamp = JSON.parse(oldTimestampStr);
+    const curTimestamp = new Date();
+    if (curTimestamp.getTime() - oldTimestamp.expiry > timeoutSeconds) {
+      localStorage.removeItem(key);
+      setTimeStamp();
+      return TutorialState.First;
+    }
+    return TutorialState.Closed;
```
Comment (AbhiPrasad) [inline ui/src/pages/ShoeMap.tsx]: If you can, please write a test to confirm that these functions work as expected.

File: ui/src/pages/ShoeMap.tsx [hunk 6]
```diff
@@ -107,7 +133,7 @@ export const ShoeMap: React.FC = () => {
   const [story, setStory] = React.useState<Story | undefined>(undefined);
   const handleOpenDrawer = (s: Story) => () => setStory(s);
   const handleCloseDrawer = () => setStory(undefined);
-  const [isTutorialOpen, setIsTutorialOpen] = useState(TutorialState.First);
+  const [isTutorialOpen, setIsTutorialOpen] = useState(compareTimestamp());
```
Comment (AbhiPrasad) [inline ui/src/pages/ShoeMap.tsx]: Nit: Can we think of a better name for `compareTimestamp()`?

File: ui/src/pages/ShoeMap.tsx [hunk 7]
```diff
@@ -89,8 +89,34 @@ const StyledLogo = styled.div`
 export const ShoeMap: React.FC = () => {
   const zoom = 4;
   const minZoom = 4;
-  const maxZoom = 18;
+  const maxZoom = 12;
+  const timeoutSeconds = 1728000000;
   const currentLocation = { lat: 53.655697, lng: -100.13316 };
+  const key = "timestamp";
```
Comment (AbhiPrasad) [inline ui/src/pages/ShoeMap.tsx]: I think we should move these constants out of the component.

Comment (AbhiPrasad) [reply ui/src/pages/ShoeMap.tsx]:
```suggestion
  const tutorialKey = "tutorial_timestamp";
```

File: ui/src/pages/ShoeMap.tsx (hunk 4 above)
Comment (hanlinc27) [reply ui/src/pages/ShoeMap.tsx]: Moved them into a separate file to make it easier to test

File: ui/src/pages/ShoeMap.tsx (hunk 5 above)
Comment (hanlinc27) [reply ui/src/pages/ShoeMap.tsx]: jesttests

File: ui/src/components/HelpAccordion.tsx [hunk 8]
```diff
@@ -89,10 +89,9 @@ export function HelpAccordion({ title, body }: HelpDrawerProps): JSX.Element {
   const [isHover, setIsHover] = useState(false);
   return (
     <StyledAccordion
-      boxShadow={0}
       onMouseEnter={() => setIsHover(true)}
       onMouseLeave={() => setIsHover(false)}
-      isHover={isHover}
+      ishover={isHover ? 1 : 0}
```
Comment (hanlinc27) [inline ui/src/components/HelpAccordion.tsx:94]: Changed b/c of console error: https://github.com/styled-components/styled-components/issues/1198

File: ui/src/components/helpers/WelcomeTutorialFunctions.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
+const tutorialKey = "tutorial_timestamp";
+
+function setTimeStamp() {
```
Comment (hanlinc27) [inline ui/src/components/helpers/WelcomeTutorialFunctions.ts]: Jest won't import it when I set return type 😞 `function setTimeStamp(): void{
`
so there is no return type in signature for tests to pass at the moment

File: ui/src/components/helpers/WelcomeTutorialFunctions.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  const dateItem = {
+    value: "value",
+    expiry: new Date().getTime(),
+  };
+  localStorage.setItem(tutorialKey, JSON.stringify(dateItem));
+}
+
+function checkIfValidTimestamp(totalTimeout) {
```
Comment (hanlinc27) [inline ui/src/components/helpers/WelcomeTutorialFunctions.ts]: Correct signature should be `function checkIfValidTimestamp(totalTimeout: number): boolean {` 
but tests give 
image

File: ui/src/components/helpers/WelcomeTutorialFunctions.ts (hunk 9 above)
Comment (AbhiPrasad) [reply ui/src/components/helpers/WelcomeTutorialFunctions.ts]: Oh hmm it's cause jest is not parsing typescript. I'll take a look tomorrow to see what's up, weird.

File: ui/src/components/helpers/WelcomeTutorialFunctions.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    return true;
+  }
+  return false;
+}
+
+const welcomeTutorialFunctions = {
+  tutorialKey,
+  setTimeStamp,
+  checkIfValidTimestamp,
+};
```
Comment (AbhiPrasad) [inline ui/src/components/helpers/WelcomeTutorialFunctions.ts]: I don't think we need to put this into an object, I'd rather just export all these individually.

File: ui/src/components/helpers/WelcomeTutorialFunctions.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
+const tutorialKey = "tutorial_timestamp";
```
Comment (AbhiPrasad) [inline ui/src/components/helpers/WelcomeTutorialFunctions.ts]: We don't need to capitalize files if it's not a react component, so 
```diff
+ ui/src/components/helpers/welcomeTutorialFunctions.ts 
- ui/src/components/helpers/WelcomeTutorialFunctions.ts 
```

File: ui/src/pages/ShoeMap.tsx [hunk 13]
```diff
@@ -107,7 +114,11 @@ export const ShoeMap: React.FC = () => {
   const [story, setStory] = React.useState<Story | undefined>(undefined);
   const handleOpenDrawer = (s: Story) => () => setStory(s);
   const handleCloseDrawer = () => setStory(undefined);
-  const [isTutorialOpen, setIsTutorialOpen] = useState(TutorialState.First);
+  const [isTutorialOpen, setIsTutorialOpen] = useState(
+    welcomeTutorialFunctions.checkIfValidTimestamp(TIMEOUT_SECONDS)
```
Comment (AbhiPrasad) [inline ui/src/pages/ShoeMap.tsx]: `checkIfValidTimestamp` -> `isTimestampExpired`?

**Review by AbhiPrasad — COMMENTED** [review]
Lemme try fix the jest stuff, sorry about that, must just be some funky stuff with parcel.

File: ui/src/__tests__/App.test.tsx [hunk 14]
```diff
@@ -1,5 +1,50 @@
-describe("<App />", () => {
-  test("test", async () => {
-    expect(1 + 1).toEqual(2);
+import {} from "jest";
```
Comment (AbhiPrasad) [inline ui/src/__tests__/App.test.tsx]: We shouldn't need the jest import

File: ui/src/components/helpers/WelcomeTutorialFunctions.ts (hunk 9 above)
Comment (hanlinc27) [reply ui/src/components/helpers/WelcomeTutorialFunctions.ts]: Yeah, it's not able to find the Jest typings either 🤷🏻 https://github.com/kulshekhar/ts-jest/issues/889 I looked into it a bit but didn't get too far

**Comment by hanlinc27** [comment]
building for production seems to be fine locally?
image
but breaking on github actions

**Comment by hanlinc27** [comment]
@AbhiPrasad should be fixed now and thanks for the jest help 😄

File: ui/package.json [hunk 15]
```diff
@@ -38,12 +39,12 @@
     "swr": "^0.3.5"
   },
   "devDependencies": {
-    "@babel/core": "^7.11.6",
+    "@babel/core": "^7.0.0-0",
```
Comment (AbhiPrasad) [inline ui/package.json]: Think we can revert the `@babel/core` and `@types/jest` version changes

File: ui/.tsconfig.json [hunk 16]
```diff
@@ -12,7 +12,8 @@
         "jsx": "react",
         "esModuleInterop": true,
         "allowSyntheticDefaultImports": true,
-        "types": ["node"]
+        "types": ["jest","node"],
+        "typeRoots": ["./node_modules/@types"]
```
Comment (AbhiPrasad) [inline ui/.tsconfig.json]: We can remove `typeRoots`

**Review by AbhiPrasad — APPROVED** [review]
Nice work, thanks for sticking with it.
